python/mxnet/text/embedding.py

```diff
@@ -0,0 +1,722 @@
+                len(new_vectors.shape) in {1, 2}, \
+                '`new_vectors` must be a 1-D or 2-D NDArray if `tokens` is a ' \
+                'singleton.'
+            if not isinstance(tokens, list):
+                tokens = [tokens]
+            if len(new_vectors.shape) == 1:
+                new_vectors = new_vectors.expand_dims(0)
+
+        else:
+            assert isinstance(tokens, list), \
+                '`tokens` must be a string or a list of strings'
+            assert isinstance(new_vectors, nd.NDArray) and \
+                len(new_vectors.shape) == 2, \
+                '`new_vectors` must be a 2-D NDArray if `tokens` is a list ' \
+                'of multiple strings.'
+        assert new_vectors.shape[0] == len(tokens), \
+            'The length of new_vectors must be equal to the number of tokens.'
+        assert new_vectors.shape[1] == self.vec_len, \
+            'The width of new_vectors must be equal to the dimension of ' \
+            'embeddings of the glossary.'
+
+        indices = []
+        for token in tokens:
+            if token in self.token_to_idx:
+                indices.append(self.token_to_idx[token])
+            else:
+                raise ValueError('Token %s is unknown. To update the embedding '
+                                 'vector for an unknown token, please specify '
+                                 'it explicitly as the `unknown_token` %s in '
+                                 '`tokens`. This is to avoid unintended '
+                                 'updates.' %
+                                 (token, self.idx_to_token[self.unknown_idx]))
+
+        self._idx_to_vec[nd.array(indices)] = new_vectors
+
+    @staticmethod
+    def register(embed_cls):
+        """Registers a new text embedding.
+
+        Once an embedding is registered, we can create an instance of this
+        embedding with :func:`~mxnet.text.embedding.TextEmbed.create`.
+
+
+        Examples
+        --------
+        >>> @mxnet.text.embedding.TextEmbed.register
+        ... class MyTextEmbed(mxnet.text.glossary.TextEmbed):
+        ...     def __init__(self, pretrain_file='my_pretrain_file'):
+        ...         pass
+        >>> embed = mxnet.text.embedding.TextEmbed.create('MyTextEmbed')
+        >>> print(type(embed))
+        <class '__main__.MyTextEmbed'>
+        """
+
+        assert(isinstance(embed_cls, type))
+        embed_name = embed_cls.__name__.lower()
+        if embed_name in TextEmbed.embed_registry:
+            warnings.warn('New embedding %s.%s is overriding existing '
+                          'embedding %s.%s', embed_cls.__module__,
+                          embed_cls.__name__,
+                          TextEmbed.embed_registry[embed_name].__module__,
+                          TextEmbed.embed_registry[embed_name].__name__)
+        TextEmbed.embed_registry[embed_name] = embed_cls
+        return embed_cls
+
+    @staticmethod
+    def create(embed_name, **kwargs):
```

Comment:
Looks like this is following what was done in optimizer module. I think both embedding and optimizer should reuse `mx.registry` [here](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/registry.py) instead of creating a new one. See example in [mx.initializer](https://github.com/apache/incubator-mxnet/blob/251ae71a20d8318ab20f4c19520f74b881fdf3ff/python/mxnet/initializer.py#L246-L248)

Reply:
Thanks. Since using `mx.registry` will make the affected code access a private member via `registry._REGISTRY[TextEmbedding]`, I will keep what we have for now. After registry allows accessing `registry._REGISTRY` publicly, we can change later.

Reply:
resolved